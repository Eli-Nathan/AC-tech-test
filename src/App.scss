.vehicleCard {
  img {
    width: 100%
  }
}
