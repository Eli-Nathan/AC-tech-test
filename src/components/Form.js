import React, { Component } from 'react'

class Form extends Component {
  constructor(props) {
    super(props);
    // Initialise all states
    this.state = {
      vehiclePrice: null,
      deposit: null,
      date: null,
      formSubmitted: false,
      arrangementFee: 88,
      completionFee: 20,
      termLength: 1,
      depositError: false,
      data: []
    }
  }

  render() {
    return (
      <div className="sm:ch-col--10 sm:ch-col--offset-1 md:ch-col--8 md:ch-col--offset-2 lg:ch-col--6 lg:ch-col--offset-3">
        <div className="ch-bg--white ch-ba--1 ch-bc--grey-2 ch-pa--3 ch-rounded">
          <div className="ch-form__group">
            <label
              htmlFor="price"
              className="ch-form__control-label">
              Vehicle Price (£)
            </label>
            <input
              className="ch-form__control"
              type="number"
              id="price" />
          </div>
          <div className={`ch-form__group${this.props.depositError && " ch-form__group--error" }`}>
            <label
              htmlFor="deposit"
              className="ch-form__control-label">
              Desposit amount (£)
            </label>
            <span className="ch-form__control-validation">{`Deposit must be at least 15% of the vehicle price, i.e. £${(15 / 100 ) * this.props.price}`}</span>
            <input
              className="ch-form__control"
              type="number"
              id="deposit" />
          </div>
          <div className="ch-form__group">
            <label
              htmlFor="deliveryDate"
              className="ch-form__control-label">
              Delivery date
            </label>
            <input
              className="ch-form__control"
              type="date"
              id="deliveryDate" />
          </div>
          <div className="ch-form__group ch-display--inline-block ch-mr--2">
            <label
              htmlFor="arrangement"
              className="ch-form__control-label">
              Arrangement fee (£)
            </label>
            <input
              className="ch-form__control"
              type="number"
              id="arrangement"
              defaultValue={88} />
          </div>
          <div className="ch-form__group ch-display--inline-block">
            <label
              htmlFor="completion"
              className="ch-form__control-label">
              Completion fee (£)
            </label>
            <input
              className="ch-form__control"
              type="number"
              id="completion"
              defaultValue={20} />
          </div>

          <div className="ch-form__group">
            <h5 className="ch-mb--1">Finance term</h5>
            <input
              id="oneYear"
              type="radio"
              name="financeTerm"
              value={1}
              className="ch-reader ch-radio"
              onChange={() => this.props.switch(1)}
              defaultChecked />
            <label
              htmlFor="oneYear"
              className="ch-display--block sm:ch-display--inline ch-radio__label ch-mb--1 ch-radio__label--compact">
              1 year
            </label>

            <input
              id="twoYears"
              type="radio"
              name="financeTerm"
              value={2}
              className="ch-reader ch-radio"
              onChange={() => this.props.switch(2)} />
            <label
              htmlFor="twoYears"
              className="ch-display--block sm:ch-display--inline sm:ch-ml--2 ch-mb--1 ch-radio__label ch-radio__label--compact">
              2 years
            </label>

            <input
              id="threeYears"
              type="radio"
              name="financeTerm"
              value={3}
              className="ch-reader ch-radio"
              onChange={() => this.props.switch(3)} />
            <label
              htmlFor="threeYears"
              className="ch-display--block sm:ch-display--inline sm:ch-ml--2 ch-mb--1 ch-radio__label ch-radio__label--compact">
              3 years
            </label>
          </div>
          <div className="ch-form__group">
            <button
              id="submit"
              className="ch-btn ch-btn--success"
              onClick={this.props.submit}>
              Calculate
            </button>
          </div>
        </div>
      </div>
    )
  }
}

export default Form
