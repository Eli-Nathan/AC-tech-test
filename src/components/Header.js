import React, { Component } from 'react'
import logo from '../ac-logo.svg'

class Header extends Component {

  render() {
    return (
      <div className="ch-text--center">
        <img
          src={logo}
          alt="logo" />
        <h1 className="ch-mt--3">{this.props.heading}</h1>
      </div>
    )
  }
}

export default Header
