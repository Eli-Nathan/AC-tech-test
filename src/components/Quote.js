import React, { Component } from 'react'

class Quote extends Component {

  render() {
    return (
      <div>
        <h3 className="ch-mt--4">Your quote</h3>
        <p>{`For your interest free loan, you will borrow £${this.props.price - this.props.deposit} over ${this.props.termLength} years. Your first payment will be due on the first Monday of the month following the delivery date. The first payment listed below is inclusive of an £${this.props.arrangementFee} arrangement fee and the final payment is inclusive of a £${this.props.completionFee}`} completion fee</p>
      </div>
    )
  }
}

export default Quote
