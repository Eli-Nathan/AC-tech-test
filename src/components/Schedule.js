import React, { Component } from 'react';
import moment from 'moment'

class Schedule extends Component {
  getMondays = (date, nextMonday) => {
    let allPaymentDates = []
    let eachDate
    let monthsToPay = this.props.termLength*12
    for(let i = 0; i < monthsToPay; i++) {
      eachDate = new Date(date.getFullYear(), (date.getMonth() + (i + 1)))
      eachDate = new Date(eachDate.setDate(eachDate.getDate() + (nextMonday+(7 - eachDate.getDay())) % 7))
      allPaymentDates.push(
        {
          key: i+1,
          date: eachDate,
          payment: this.props.monthlyMax
        }
      )
    }
    // Add 88 initial payment to first date
    allPaymentDates[0].payment = allPaymentDates[0].payment + this.props.arrangementFee;
    // Add 20 settlement payment to last date
    allPaymentDates[allPaymentDates.length -1].payment = allPaymentDates[allPaymentDates.length -1].payment + this.props.completionFee;
    return allPaymentDates
  }

  renderSchedule = () => {
    const allDates = this.getMondays(this.props.date, 1).map(payment => (
      <tr key={payment.key}>
        <td>{moment(payment.date).format("Do MMM YYYY")}</td>
        <td>Â£{payment.payment.toFixed(2)}</td>
      </tr>
    ))
    return allDates
  }

  render() {
    return (
      <div>
        <h3 className="ch-mt--4">Payment Schedule</h3>
        <table
          className="ch-table ch-table--bordered ch-table--hover ch-table--striped ch-mt--2"
          width="100%">
          <thead>
            <tr>
              <th>Date</th>
              <th>Amount payable</th>
            </tr>
          </thead>
          <tbody>
            {this.renderSchedule()}
          </tbody>
        </table>
      </div>
    );
  }
}

export default Schedule;
