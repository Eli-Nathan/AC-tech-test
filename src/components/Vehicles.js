import React, { Component } from 'react'

class Vehicles extends Component {
  constructor(props) {
    super(props)
    this.renderVehicles = this.renderVehicles.bind(this)
  }

  renderVehicles = () => {
    let vehicles = "Sorry, no vehicles found"
    let data = this.props.data.filter(d => d.salesInfo.pricing.monthlyPayment <= this.props.monthlyMax)
    if(data.length === 0) {
      return vehicles
    }

    if(data.length > 6) {
      data.length = 6
    }
    return data.map(vehicle => (
      <div className="sm:ch-col--6 md:ch-col--4 ch-mb--2 vehicleCard" key={vehicle.stockReference}>
        <div className="ch-bg--white ch-ba--1 ch-bc--grey-2 ch-rounded">
          <a href={vehicle.url} className="ch-color--ac-black ch-text-decoration--none">
            <div>
              <img
                src={vehicle.photos.length === 0 ? "https://www.arnoldclark.com/assets/application/no-images-vehicle-8c4df08470af32dee8f1a3ea091e72cbf2bac4225054b1844de0880d1bf766ae.svg" : vehicle.photos[0]}
                alt={`${vehicle.make} ${vehicle.model}`}
                className="ch-img-responsive "/>
            </div>
            <div className="ch-pa--2">
              <h4>{vehicle.title.name}</h4>
              <h5 className="ch-color--grey-5">{vehicle.title.variant}</h5>
            </div>
          </a>
        </div>
      </div>
    ))
  }

  render() {
    return (
      <div>
        <div className="sm:ch-col--12">
          <h3 className="ch-mt--4">Recommended vehicles</h3>
        </div>
        {this.renderVehicles()}
      </div>
    )
  }
}

export default Vehicles
