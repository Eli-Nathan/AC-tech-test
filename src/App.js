import React, { Component } from 'react';
import logo from './ac-logo.svg';
import './App.scss';
import Form from './components/Form.js';

class App extends Component {

  render() {
    return (
      <div className="App">
        <header className="App-header ch-text--center">
          <img src={logo} className="App-logo" alt="logo" />
          <h1 className="ch-mt--3">Interest free loan calculator</h1>
        </header>
        <div className="sm:ch-col--10 sm:ch-col--offset-1 md:ch-col--8 md:ch-col--offset-2 lg:ch-col--6 lg:ch-col--offset-3">
          <Form />
        </div>
      </div>
    );
  }
}

export default App;
