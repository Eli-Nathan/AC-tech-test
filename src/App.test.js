import React from 'react'
import ReactDOM from 'react-dom'
import { shallow, mount, render, configure } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import App from './App'
import Header from './components/Header'
import Form from './components/Form'
import Quote from './components/Quote'
import Schedule from './components/Schedule'
import Vehicles from './components/Vehicles'

configure({ adapter: new Adapter() });

describe('<App />', () => {
  it('Component renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<App />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('matches the snapshot', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
})

describe('<Header />', () => {
  it('Successfully render header', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Header />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('matches the snapshot', () => {
    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });
})

describe('<Form />', () => {
  it('Successfully render form', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Form />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  // it('autosets date', () => {
  //   const props = {
  //     date: new Date(2019-02-24T18:36:25.175Z)
  //   },
  //   date = mount(<Form {...props} />);
  //   expect((date).prop('date')).toEqual(new Date(2019-02-24T18:36:25.175Z));
  // });

  it('matches the snapshot', () => {
    const tree = renderer.create(<Form />).toJSON();
    expect(tree).toMatchSnapshot();
  });
})

describe('<Quote />', () => {
  it('Successfully render quote', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Quote />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('matches the snapshot', () => {
    const tree = renderer.create(<Quote />).toJSON();
    expect(tree).toMatchSnapshot();
  });
})

describe('<Schedule />', () => {
  it('Successfully render schedule', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <Schedule
        date={new Date()}
        termLength={1}
        monthlyMax={150}
        arrangementFee={88}
        completionFee={20} />, div
    )
    ReactDOM.unmountComponentAtNode(div)
  })

  it('matches the snapshot', () => {
    const tree = renderer.create(<Schedule
      date={new Date()}
      termLength={1}
      monthlyMax={150}
      arrangementFee={88}
      completionFee={20} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
})

describe('<Vehicles />', () => {
  it('Successfully render vehicles', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Vehicles data={[]} monthlyMax={200} />, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('matches the snapshot', () => {
    const tree = renderer.create(<Vehicles data={[]} monthlyMax={200} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
})

it('Index finds root element to render app', () => {
  const div = document.createElement('div')
  ReactDOM.render(<App/>, div)
  global.document.getElementById = (id) => id ==='root' && div
})
