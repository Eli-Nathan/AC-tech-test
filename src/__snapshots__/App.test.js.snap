// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches the snapshot 1`] = `
<div
  className="ch-pv--4"
>
  <div
    className="ch-text--center"
  >
    <img
      alt="logo"
      src="ac-logo.svg"
    />
    <h1
      className="ch-mt--3"
    >
      Interest free loan calculator
    </h1>
  </div>
  <div
    className="sm:ch-col--10 sm:ch-col--offset-1 md:ch-col--8 md:ch-col--offset-2 lg:ch-col--6 lg:ch-col--offset-3"
  >
    <div
      className="ch-bg--white ch-ba--1 ch-bc--grey-2 ch-pa--3 ch-rounded"
    >
      <div
        className="ch-form__group"
      >
        <label
          className="ch-form__control-label"
          htmlFor="price"
        >
          Vehicle Price (£)
        </label>
        <input
          className="ch-form__control"
          id="price"
          type="number"
        />
      </div>
      <div
        className="ch-form__groupfalse"
      >
        <label
          className="ch-form__control-label"
          htmlFor="deposit"
        >
          Desposit amount (£)
        </label>
        <span
          className="ch-form__control-validation"
        >
          Deposit must be at least 15% of the vehicle price, i.e. £NaN
        </span>
        <input
          className="ch-form__control"
          id="deposit"
          type="number"
        />
      </div>
      <div
        className="ch-form__group"
      >
        <label
          className="ch-form__control-label"
          htmlFor="deliveryDate"
        >
          Delivery date
        </label>
        <input
          className="ch-form__control"
          id="deliveryDate"
          type="date"
        />
      </div>
      <div
        className="ch-form__group ch-display--inline-block ch-mr--2"
      >
        <label
          className="ch-form__control-label"
          htmlFor="arrangement"
        >
          Arrangement fee (£)
        </label>
        <input
          className="ch-form__control"
          defaultValue={88}
          id="arrangement"
          type="number"
        />
      </div>
      <div
        className="ch-form__group ch-display--inline-block"
      >
        <label
          className="ch-form__control-label"
          htmlFor="completion"
        >
          Completion fee (£)
        </label>
        <input
          className="ch-form__control"
          defaultValue={20}
          id="completion"
          type="number"
        />
      </div>
      <div
        className="ch-form__group"
      >
        <h5
          className="ch-mb--1"
        >
          Finance term
        </h5>
        <input
          className="ch-reader ch-radio"
          defaultChecked={true}
          id="oneYear"
          name="financeTerm"
          onChange={[Function]}
          type="radio"
          value={1}
        />
        <label
          className="ch-display--block sm:ch-display--inline ch-radio__label ch-mb--1 ch-radio__label--compact"
          htmlFor="oneYear"
        >
          1 year
        </label>
        <input
          className="ch-reader ch-radio"
          id="twoYears"
          name="financeTerm"
          onChange={[Function]}
          type="radio"
          value={2}
        />
        <label
          className="ch-display--block sm:ch-display--inline sm:ch-ml--2 ch-mb--1 ch-radio__label ch-radio__label--compact"
          htmlFor="twoYears"
        >
          2 years
        </label>
        <input
          className="ch-reader ch-radio"
          id="threeYears"
          name="financeTerm"
          onChange={[Function]}
          type="radio"
          value={3}
        />
        <label
          className="ch-display--block sm:ch-display--inline sm:ch-ml--2 ch-mb--1 ch-radio__label ch-radio__label--compact"
          htmlFor="threeYears"
        >
          3 years
        </label>
      </div>
      <div
        className="ch-form__group"
      >
        <button
          className="ch-btn ch-btn--success"
          id="submit"
          onClick={[Function]}
        >
          Calculate
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Form /> matches the snapshot 1`] = `
<div
  className="sm:ch-col--10 sm:ch-col--offset-1 md:ch-col--8 md:ch-col--offset-2 lg:ch-col--6 lg:ch-col--offset-3"
>
  <div
    className="ch-bg--white ch-ba--1 ch-bc--grey-2 ch-pa--3 ch-rounded"
  >
    <div
      className="ch-form__group"
    >
      <label
        className="ch-form__control-label"
        htmlFor="price"
      >
        Vehicle Price (£)
      </label>
      <input
        className="ch-form__control"
        id="price"
        type="number"
      />
    </div>
    <div
      className="ch-form__groupundefined"
    >
      <label
        className="ch-form__control-label"
        htmlFor="deposit"
      >
        Desposit amount (£)
      </label>
      <span
        className="ch-form__control-validation"
      >
        Deposit must be at least 15% of the vehicle price, i.e. £NaN
      </span>
      <input
        className="ch-form__control"
        id="deposit"
        type="number"
      />
    </div>
    <div
      className="ch-form__group"
    >
      <label
        className="ch-form__control-label"
        htmlFor="deliveryDate"
      >
        Delivery date
      </label>
      <input
        className="ch-form__control"
        id="deliveryDate"
        type="date"
      />
    </div>
    <div
      className="ch-form__group ch-display--inline-block ch-mr--2"
    >
      <label
        className="ch-form__control-label"
        htmlFor="arrangement"
      >
        Arrangement fee (£)
      </label>
      <input
        className="ch-form__control"
        defaultValue={88}
        id="arrangement"
        type="number"
      />
    </div>
    <div
      className="ch-form__group ch-display--inline-block"
    >
      <label
        className="ch-form__control-label"
        htmlFor="completion"
      >
        Completion fee (£)
      </label>
      <input
        className="ch-form__control"
        defaultValue={20}
        id="completion"
        type="number"
      />
    </div>
    <div
      className="ch-form__group"
    >
      <h5
        className="ch-mb--1"
      >
        Finance term
      </h5>
      <input
        className="ch-reader ch-radio"
        defaultChecked={true}
        id="oneYear"
        name="financeTerm"
        onChange={[Function]}
        type="radio"
        value={1}
      />
      <label
        className="ch-display--block sm:ch-display--inline ch-radio__label ch-mb--1 ch-radio__label--compact"
        htmlFor="oneYear"
      >
        1 year
      </label>
      <input
        className="ch-reader ch-radio"
        id="twoYears"
        name="financeTerm"
        onChange={[Function]}
        type="radio"
        value={2}
      />
      <label
        className="ch-display--block sm:ch-display--inline sm:ch-ml--2 ch-mb--1 ch-radio__label ch-radio__label--compact"
        htmlFor="twoYears"
      >
        2 years
      </label>
      <input
        className="ch-reader ch-radio"
        id="threeYears"
        name="financeTerm"
        onChange={[Function]}
        type="radio"
        value={3}
      />
      <label
        className="ch-display--block sm:ch-display--inline sm:ch-ml--2 ch-mb--1 ch-radio__label ch-radio__label--compact"
        htmlFor="threeYears"
      >
        3 years
      </label>
    </div>
    <div
      className="ch-form__group"
    >
      <button
        className="ch-btn ch-btn--success"
        id="submit"
      >
        Calculate
      </button>
    </div>
  </div>
</div>
`;

exports[`<Header /> matches the snapshot 1`] = `
<div
  className="ch-text--center"
>
  <img
    alt="logo"
    src="ac-logo.svg"
  />
  <h1
    className="ch-mt--3"
  />
</div>
`;

exports[`<Quote /> matches the snapshot 1`] = `
<div>
  <h3
    className="ch-mt--4"
  >
    Your quote
  </h3>
  <p>
    For your interest free loan, you will borrow £NaN over undefined years. Your first payment will be due on the first Monday of the month following the delivery date. The first payment listed below is inclusive of an £undefined and the final payment is inclusive of a £undefined
  </p>
</div>
`;

exports[`<Schedule /> matches the snapshot 1`] = `
<div>
  <h3
    className="ch-mt--4"
  >
    Payment Schedule
  </h3>
  <table
    className="ch-table ch-table--bordered ch-table--hover ch-table--striped ch-mt--2"
    width="100%"
  >
    <thead>
      <tr>
        <th>
          Date
        </th>
        <th>
          Amount payable
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          4th Mar 2019
        </td>
        <td>
          £
          238.00
        </td>
      </tr>
      <tr>
        <td>
          1st Apr 2019
        </td>
        <td>
          £
          150.00
        </td>
      </tr>
      <tr>
        <td>
          6th May 2019
        </td>
        <td>
          £
          150.00
        </td>
      </tr>
      <tr>
        <td>
          3rd Jun 2019
        </td>
        <td>
          £
          150.00
        </td>
      </tr>
      <tr>
        <td>
          1st Jul 2019
        </td>
        <td>
          £
          150.00
        </td>
      </tr>
      <tr>
        <td>
          5th Aug 2019
        </td>
        <td>
          £
          150.00
        </td>
      </tr>
      <tr>
        <td>
          2nd Sep 2019
        </td>
        <td>
          £
          150.00
        </td>
      </tr>
      <tr>
        <td>
          7th Oct 2019
        </td>
        <td>
          £
          150.00
        </td>
      </tr>
      <tr>
        <td>
          4th Nov 2019
        </td>
        <td>
          £
          150.00
        </td>
      </tr>
      <tr>
        <td>
          2nd Dec 2019
        </td>
        <td>
          £
          150.00
        </td>
      </tr>
      <tr>
        <td>
          6th Jan 2020
        </td>
        <td>
          £
          150.00
        </td>
      </tr>
      <tr>
        <td>
          3rd Feb 2020
        </td>
        <td>
          £
          170.00
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`<Vehicles /> matches the snapshot 1`] = `
<div>
  <div
    className="sm:ch-col--12"
  >
    <h3
      className="ch-mt--4"
    >
      Recommended vehicles
    </h3>
  </div>
  Sorry, no vehicles found
</div>
`;
